Use GameController API to determine mouse pointer availability instead of BKSMousePointerService

rr-codes opened the pull request:
#### 7fe4d93de5c66c57bb0994b1699b212e542fef6c
<pre>
Use GameController API to determine mouse pointer availability instead of BKSMousePointerService
<a href="https://bugs.webkit.org/show_bug.cgi?id=294709">https://bugs.webkit.org/show_bug.cgi?id=294709</a>
<a href="https://rdar.apple.com/153328349">rdar://153328349</a>

Reviewed by Aditya Keerthi.

GameController offers API to determine mouse pointer availability through notifications, so switch to using that for
the implementation of WKMouseDeviceObserver instead of BackBoard.

Consequently, remove a lot of SPI forward declarations.

* Source/WebKit/Modules/Internal/WebKitInternal.h:
* Source/WebKit/Platform/spi/ios/BackBoardServicesSPI.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/WKMouseDeviceObserver.swift: Added.
(WKMouseDeviceObserver.connectionObservationTask):
(WKMouseDeviceObserver.disconnectionObservationTask):
(WKMouseDeviceObserver.hasMouseDevice):
(WKMouseDeviceObserver.sharedInstance):
(WKMouseDeviceObserver.start):
(WKMouseDeviceObserver.stop):
(WKMouseDeviceObserver._setHasMouseDevice(forTesting:)):
* Source/WebKit/UIProcess/ios/WKMouseDeviceObserver.h:
* Source/WebKit/UIProcess/ios/WKMouseDeviceObserver.mm:
(-[WKMouseDeviceObserver mousePointerDevicesDidChange:]):
(+[WKMouseDeviceObserver sharedInstance]): Deleted.
(-[WKMouseDeviceObserver init]): Deleted.
(-[WKMouseDeviceObserver start]): Deleted.
(-[WKMouseDeviceObserver startWithCompletionHandler:]): Deleted.
(-[WKMouseDeviceObserver stop]): Deleted.
(-[WKMouseDeviceObserver stopWithCompletionHandler:]): Deleted.
(-[WKMouseDeviceObserver _setHasMouseDeviceForTesting:]): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/296457@main">https://commits.webkit.org/296457@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/62c3ad46639a09d4c116b6d6b1c517e82ba558f2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108593 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28254 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18678 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113801 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58983 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110556 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28943 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36808 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82480 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111541 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22972 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97812 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62917 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22388 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15951 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58515 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92341 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16002 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116922 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35647 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26280 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91502 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36020 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94083 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91305 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23258 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36205 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13966 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35548 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41081 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35258 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38604 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36936 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/Modules/Internal/WebKitInternal.h:
```diff
@@ -35,3 +35,4 @@
 #import "WKUIDelegateInternal.h"
 #import "WKWebViewConfigurationInternal.h"
 #import "WKWebViewInternal.h"
+#import "WKMouseDeviceObserver.h"
```
fix ordering

rr-codes commented on Source/WebKit/UIProcess/WKMouseDeviceObserver.swift:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    func start() {
+        NSLog("RR_CODES start")
```
remove

rr-codes commented on Source/WebKit/UIProcess/WKMouseDeviceObserver.swift:
```diff
@@ -0,0 +1,101 @@
+    private var startCount = 0
+    private var connectedDeviceCount = 0 {
+        didSet {
+            NSLog("RR_CODES connectedDeviceCount \(connectedDeviceCount)")
```
remove

pxlcoder commented on Source/WebKit/UIProcess/WKMouseDeviceObserver.swift:
```diff
@@ -0,0 +1,101 @@
+        }
+        startCount -= 1
+
+        connectionObservationTask?.cancel()
+        disconnectionObservationTask?.cancel()
```
Shouldn't we only cancel once the start count hits zero?

pxlcoder commented on Source/WebKit/UIProcess/WKMouseDeviceObserver.swift:
```diff
@@ -0,0 +1,101 @@
+        guard startCount > 0 else {
+            return
+        }
+        startCount -= 1
```
```suggestion

        startCount -= 1
```

pxlcoder commented on Source/WebKit/UIProcess/WKMouseDeviceObserver.swift:
```diff
@@ -0,0 +1,101 @@
+#if HAVE_MOUSE_DEVICE_OBSERVATION
+
+internal import WebKit_Internal
+internal import GameController
```
Sort order?

pxlcoder commented on Source/WebKit/UIProcess/WKMouseDeviceObserver.swift:
```diff
@@ -0,0 +1,103 @@
+            precondition(connectedDeviceCount >= 0)
+
+            if oldValue == 0 && connectedDeviceCount > 0 {
+                mousePointerDevicesDidChange(true)
```
We can early return after this line. No need to check the other condition.

rr-codes replied:
fixed!

pxlcoder commented on Source/WebKit/UIProcess/WKMouseDeviceObserver.swift:
```diff
@@ -0,0 +1,106 @@
+            return
+        }
+
+        connectionObservationTask = Task {
```
Does this need to weakly capture self to avoid a retain cycle?

rr-codes replied:
it does not, since the `stop` function cancels the task.

pxlcoder commented on Source/WebKit/UIProcess/WKMouseDeviceObserver.swift:
```diff
@@ -0,0 +1,106 @@
+
+            if oldValue > 0 && connectedDeviceCount == 0 {
+                mousePointerDevicesDidChange(false)
+                return
```
This return is unnecessary.

rr-codes committed:
Use GameController API to determine mouse pointer availability instead of BKSMousePointerService
https://bugs.webkit.org/show_bug.cgi?id=294709
rdar://153328349

Reviewed by Aditya Keerthi.

GameController offers API to determine mouse pointer availability through notifications, so switch to using that for
the implementation of WKMouseDeviceObserver instead of BackBoard.

Consequently, remove a lot of SPI forward declarations.

* Source/WebKit/Modules/Internal/WebKitInternal.h:
* Source/WebKit/Platform/spi/ios/BackBoardServicesSPI.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/WKMouseDeviceObserver.swift: Added.
(WKMouseDeviceObserver.connectionObservationTask):
(WKMouseDeviceObserver.disconnectionObservationTask):
(WKMouseDeviceObserver.hasMouseDevice):
(WKMouseDeviceObserver.sharedInstance):
(WKMouseDeviceObserver.start):
(WKMouseDeviceObserver.stop):
(WKMouseDeviceObserver._setHasMouseDevice(forTesting:)):
* Source/WebKit/UIProcess/ios/WKMouseDeviceObserver.h:
* Source/WebKit/UIProcess/ios/WKMouseDeviceObserver.mm:
(-[WKMouseDeviceObserver mousePointerDevicesDidChange:]):
(+[WKMouseDeviceObserver sharedInstance]): Deleted.
(-[WKMouseDeviceObserver init]): Deleted.
(-[WKMouseDeviceObserver start]): Deleted.
(-[WKMouseDeviceObserver startWithCompletionHandler:]): Deleted.
(-[WKMouseDeviceObserver stop]): Deleted.
(-[WKMouseDeviceObserver stopWithCompletionHandler:]): Deleted.
(-[WKMouseDeviceObserver _setHasMouseDeviceForTesting:]): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/296457@main
